src/libqof/qof/gnc-numeric.cpp

```diff
@@ -21,46 +21,501 @@
+        m_num = n.num();
+        m_den = n.denom();
+        return;
+    }
+    if (regex_search(str, m, numeral))
+    {
+        GncNumeric n(stoll(m[1].str()), INT64_C(1));
+        m_num = n.num();
+        m_den = n.denom();
+        return;
+    }
+    std::ostringstream errmsg;
+    errmsg << "String " << str << " contains no recognizable numeric value.";
+    throw std::invalid_argument(errmsg.str());
+}
+
+GncNumeric::operator gnc_numeric() const noexcept
+{
+    return {m_num, m_den};
+}
+
+GncNumeric::operator double() const noexcept
+{
+    return static_cast<double>(m_num) / static_cast<double>(m_den);
+}
+
+GncNumeric
+GncNumeric::operator-() const noexcept
+{
+    GncNumeric b(*this);
+    b.m_num = - b.m_num;
+    return b;
+}
+
+GncNumeric
+GncNumeric::inv() const noexcept
 {
-    if (exp > 18 || exp < -18)
-	return POWTEN_OVERFLOW;
-    return exp < 0 ? -pten[-exp] : pten[exp];
+    if (m_num == 0)
+        return *this;
+    if (m_num < 0)
+        return GncNumeric(-m_den, -m_num);
+    return GncNumeric(m_den, m_num);
 }
 
+GncNumeric
+GncNumeric::abs() const noexcept
+{
+    if (m_num < 0)
+        return -*this;
+    return *this;
+}
+
+GncNumeric
+GncNumeric::reduce() const noexcept
+{
+    return static_cast<GncNumeric>(GncRational(*this).reduce());
+}
+
+GncNumeric::round_param
+GncNumeric::prepare_conversion(int64_t new_denom) const
+{
+    if (new_denom == m_den || new_denom == GNC_DENOM_AUTO)
+        return {m_num, m_den, 0};
+    GncRational conversion(new_denom, m_den);
+    auto red_conv = conversion.reduce();
+    GncInt128 old_num(m_num);
+    auto new_num = old_num * red_conv.num();
+    auto rem = new_num % red_conv.denom();
+    new_num /= red_conv.denom();
+    if (new_num.isBig())
+    {
+        GncRational rr(new_num, new_denom);
+        GncNumeric nn(rr);
+        rr = rr.convert<RoundType::truncate>(new_denom);
+        return {static_cast<int64_t>(rr.num()), new_denom, 0};
+    }
+    return {static_cast<int64_t>(new_num),
+            static_cast<int64_t>(red_conv.denom()), static_cast<int64_t>(rem)};
+}
+
+int64_t
+GncNumeric::sigfigs_denom(unsigned figs) const noexcept
+{
+    int64_t num_abs{std::abs(m_num)};
+    bool not_frac = num_abs > m_den;
+    int64_t val{ not_frac ? num_abs / m_den : m_den / num_abs };
+    unsigned digits{};
+    while (val >= 10)
+    {
+        ++digits;
+        val /= 10;
+    }
+    return not_frac ? powten(figs - digits - 1) : powten(figs + digits);
+}
+
+std::string
+GncNumeric::to_string() const noexcept
+{
+    std::ostringstream out;
+    out << *this;
+    return out.str();
+}
+
+GncNumeric
+GncNumeric::to_decimal(unsigned int max_places) const
+{
+    if (max_places > 17)
+        max_places = 17;
+    bool is_pwr_ten = true;
+    for (int pwr = 1; pwr <= 17 && m_den > powten(pwr); ++pwr)
+        if (m_den % powten(pwr))
+        {
+            is_pwr_ten = false;
+            break;
+        }
+
```

Comment:
It looks like this will give invalid results if m_denom < 10.
In that case the for loop is never executed, yet the GncNumeric is not necessarily a power of ten. I discovered this by chance because my data holds an amount of 21/2. This will eventually result in an Arithmetic exception being thrown in xaccPrintAmountInternal because it makes reductions that assume the denominator is really a power of ten (in the while loop starting at src/app-utils/gnc-ui-utils.c:1509).
One way to mitigate this can be to properly initialize is_pwr_ten:
`bool is_pwr_ten = (m_den >= 10);`